src/main/java/io/github/mfvanek/pg/connection/HighAvailabilityPgConnectionImpl.java

```diff
@@ -10,26 +10,40 @@
 
 package io.github.mfvanek.pg.connection;
 
+import io.github.mfvanek.pg.utils.PgSqlException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashSet;
 import java.util.Objects;
 import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
 import javax.annotation.Nonnull;
 
+@SuppressWarnings("PMD.DoNotUseThreads")
 public class HighAvailabilityPgConnectionImpl implements HighAvailabilityPgConnection {
 
-    // TODO Bad design. Need logic here to deal with failover/switch-over in real cluster.
-    //  As a possible solution - cache connectionToPrimary for a short period of time (e.g. 1 minute)
-    private final PgConnection connectionToPrimary;
+    private static final Logger LOGGER = LoggerFactory.getLogger(HighAvailabilityPgConnectionImpl.class);
+    private static final Integer DEFAULT_DELAY_SECONDS = 30;
+
+    private final AtomicReference<PgConnection> cachedConnectionToPrimary = new AtomicReference<>();
     private final Set<PgConnection> connectionsToAllHostsInCluster;
+    private final ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+    private final PrimaryHostDeterminer primaryHostDeterminer;
 
     private HighAvailabilityPgConnectionImpl(@Nonnull final PgConnection connectionToPrimary,
-                                             @Nonnull final Collection<PgConnection> connectionsToAllHostsInCluster) {
-        this.connectionToPrimary = Objects.requireNonNull(connectionToPrimary, "connectionToPrimary");
```

Comment:
It's incorrect to remove `Objects.requireNonNull(connectionToPrimary, "connectionToPrimary")` here.